presto-main/src/main/java/com/facebook/presto/type/khyperloglog/KHyperLogLog.java

```diff
@@ -0,0 +1,285 @@
+            minhash.put(input.readLong(), HyperLogLog.newInstance(serializedHll));
+        }
+
+        return new KHyperLogLog(maxSize, hllBuckets, minhash);
+    }
+
+    public static long exactIntersectionCardinality(KHyperLogLog a, KHyperLogLog b)
+    {
+        checkState(a.isExact(), "exact intersection cannot operate on approximate sets");
+        checkArgument(b.isExact(), "exact intersection cannot operate on approximate sets");
+
+        return Sets.intersection(a.minhash.keySet(), b.minhash.keySet()).size();
+    }
+
+    public static double jaccardIndex(KHyperLogLog a, KHyperLogLog b)
+    {
+        int sizeOfSmallerSet = Math.min(a.minhash.size(), b.minhash.size());
+        LongSortedSet minUnion = new LongRBTreeSet(a.minhash.keySet());
+        minUnion.addAll(b.minhash.keySet());
+
+        int intersection = 0;
+        int i = 0;
+        for (long key : minUnion) {
+            if (a.minhash.containsKey(key) && b.minhash.containsKey(key)) {
+                intersection++;
+            }
+            i++;
+            if (i >= sizeOfSmallerSet) {
+                break;
+            }
+        }
+        return intersection / (double) sizeOfSmallerSet;
+    }
+
+    public static KHyperLogLog merge(KHyperLogLog khll1, KHyperLogLog khll2)
+    {
+        /*
+        Return the one with smallest K so resolution is not lost. This loss would happen in the case
+        one merged a smaller KHLL into a bigger one because the former's minhash struct won't
+        cover all of the latter's minhash space.
+         */
+        if (khll1.maxSize <= khll2.maxSize) {
+            return khll1.mergeWith(khll2);
+        }
+        return khll2.mergeWith(khll1);
+    }
+
+    public boolean isExact()
+    {
+        return minhash.size() < maxSize;
+    }
+
+    public long getMinhashSize()
+    {
+        return minhash.size();
+    }
+
+    public Slice serialize()
+    {
+        try (SliceOutput output = new DynamicSliceOutput(estimatedSerializedSize())) {
+            output.appendByte(UNCOMPRESSED_FORMAT);
+            output.appendInt(maxSize);
+            output.appendInt(hllBuckets);
+            output.appendInt(minhash.size());
+            for (long key : minhash.keySet()) {
+                output.appendLong(key);
+            }
+            Slice serializedHll;
+            for (HyperLogLog hll : minhash.values()) {
+                serializedHll = hll.serialize();
+                output.appendInt(serializedHll.length());
+                output.appendBytes(serializedHll);
+            }
+
+            return output.slice();
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public int estimatedInMemorySize()
+    {
+        return SIZE_OF_KHYPERLOGLOG +
+                SIZE_OF_RBTREEMAP +
+                minhash.size() * SIZE_OF_LONG +
+                minhash.values().stream().mapToInt(HyperLogLog::estimatedInMemorySize).sum();
```

Comment:
As we discussed, we want to try to compute this as we go rather than in this method.

Reply:
Right, that's being implemented. Will be included in upcoming version